tracing: fix corruption of structured recording buffer

andreimatei opened the pull request:
This patch fixes a bug leading to corruption of one of the Span's
recording buffers.
We had broken code like the following:
```
for _, e := range c.StructuredRecords {
        s.recordInternalLocked(&e, &s.mu.recording.structured)
}
```
`recordInternalLocked(ptr)` takes ownership of ptr, so the `&e` is
broken because `e` keep being reassigned.

The bug was causing buffer elements to essentially be overwritten. In
turn, this was leading to inconsistencies in the accounting of the size
of the buffer's elements - we were maintaining a size sum that was
diverging from the overwritten reality.

Fixes #74994
Fixes #75045

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/75099)
<!-- Reviewable:end -->

andreimatei committed:
tracing: fix corruption of structured recording buffer

This patch fixes a bug leading to corruption of one of the Span's
recording buffers.
We had broken code like the following:
```
for _, e := range c.StructuredRecords {
        s.recordInternalLocked(&e, &s.mu.recording.structured)
}
```
`recordInternalLocked(ptr)` takes ownership of ptr, so the `&e` is
broken because `e` keep being reassigned.

The bug was causing buffer elements to essentially be overwritten. In
turn, this was leading to inconsistencies in the accounting of the size
of the buffer's elements - we were maintaining a size sum that was
diverging from the overwritten reality.

Fixes #74994
Fixes #75045

Release note: None

abarganier reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/75099#-:-Mtj53dgD7c_SmqhiwI9:bg7igkv)* status: :shipit: complete! 1 of 0 LGTMs obtained



<!-- Sent from Reviewable.io -->

irfansharif commented on pkg/util/tracing/crdbspan.go:
```diff
@@ -375,9 +378,10 @@ func (s *crdbSpan) recordFinishedChildrenLocked(children []tracingpb.RecordedSpa
-		for _, c := range children {
-			for _, e := range c.StructuredRecords {
-				s.recordInternalLocked(&e, &s.mu.recording.structured)
+		for c := range children {
```
[mega nit] `c` looks like it'd be of type tracingpb.RecordedSpan but it's an index.

andreimatei replied:
renamed to ci

irfansharif reviewed (approved):
Any idea why #73876 made it worse/was bisected to?

andreimatei commented:
> Any idea why #73876 made it worse/was bisected to?

Not really. The symptoms that the failed tests report (trying to read the first element of an empty buffer) are caused by the fact that the computed size of the buffer is not kept in sync with reality, which is a problem when a span overflows its byte limit for the structured recordings. So one thing that would trigger it is a change that makes some span accumulate more structured recordings than before.

knz reviewed (approved):
> Reviewed 2 of 2 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/75099)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @irfansharif)



<!-- Sent from Reviewable.io -->

andreimatei reviewed (commented):
bors r+

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/75099#-:-MtoFAI6BDCufRRHMalS:b-ydi9ay)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @irfansharif)



<!-- Sent from Reviewable.io -->

andreimatei commented:
bors r+
